pkg/middlewares/aws_api_gateway_domain_names_reconciler.go

```diff
@@ -0,0 +1,81 @@
+package middlewares
+
+import (
+	"github.com/snyk/driftctl/pkg/resource"
+	"github.com/snyk/driftctl/pkg/resource/aws"
+)
+
+// TODO
+type AwsApiGatewayDomainNamesReconciler struct{}
+
+func NewAwsApiGatewayDomainNamesReconciler() AwsApiGatewayDomainNamesReconciler {
+	return AwsApiGatewayDomainNamesReconciler{}
+}
+
+func (m AwsApiGatewayDomainNamesReconciler) Execute(remoteResources, resourcesFromState *[]*resource.Resource) error {
+	newRemoteResources := make([]*resource.Resource, 0)
+	managedDomainNames := make([]*resource.Resource, 0)
+	unmanagedDomainNames := make([]*resource.Resource, 0)
+	for _, res := range *remoteResources {
+		// Ignore all resources other than aws_api_gateway_domain_name and aws_apigatewayv2_domain_name
+		if res.ResourceType() != aws.AwsApiGatewayDomainNameResourceType &&
+			res.ResourceType() != aws.AwsApiGatewayV2DomainNameResourceType {
+			newRemoteResources = append(newRemoteResources, res)
+			continue
+		}
+
+		// Find a matching state resources
+		existInState := false
+		for _, stateResource := range *resourcesFromState {
+			if res.Equal(stateResource) {
+				existInState = true
+				break
+			}
+		}
+
+		// Keep track of the resource if it's managed in IaC
+		if existInState {
+			managedDomainNames = append(managedDomainNames, res)
+			continue
+		}
+
+		// If we're here, it means that we are left with unmanaged domain names
+		// in both v1 and v2 format. Let's group real and duplicate domain names
+		// in a slice
+		unmanagedDomainNames = append(unmanagedDomainNames, res)
+	}
+
+	// We only want to show to our end users unmanaged v1 domain names
+	// To do that, we're going to loop over unmanaged domain names to delete duplicates
+	// and leave after that only v1 domain names (e.g. remove v2 ones)
+	for i := 0; i < len(unmanagedDomainNames); i++ {
+		unmanaged := unmanagedDomainNames[i]
+
+		// Remove duplicates (e.g. same id, the opposite type)
+		isDuplicate := false
+		for _, managed := range managedDomainNames {
+			if managed.ResourceId() == unmanaged.ResourceId() {
+				isDuplicate = true
+				break
+			}
+		}
+		if isDuplicate {
+			unmanagedDomainNames = append(unmanagedDomainNames[:i], unmanagedDomainNames[i+1:]...)
+			i--
+			continue
+		}
+
+		// Now remove only v2 domain names
+		if unmanaged.ResourceType() == aws.AwsApiGatewayV2DomainNameResourceType {
+			unmanagedDomainNames = append(unmanagedDomainNames[:i], unmanagedDomainNames[i+1:]...)
+			i--
+		}
+	}
```

Comment:
I always feel uncomfortable by mutating array that we are currently iterating on.
I don't really know how golang handle this under the hoods but I suppose that it creates a copy of the slice 🤔

WDYT of explicitly creating a new `deduplicatedUnmanagedDomains` slice instead and appending on it while iterating on `unmanagedDomainNames` ?

I think it will lead to a code that will be easier to maintain.

@craigfurman Nevermind of this comment if your schedule is full, but I would love to have your goland expert eye on this remark 🙏🏻 I'm kinda new in the goland world and would be very happy to know more 😃
I could probably have found the answer by myself 😅